+sw_tests/+unit_tests/unittest_spinw_optmagstr.m

```diff
@@ -0,0 +1,235 @@
+            end
+            testCase.tri.optmagstr('func', @custom_func, 'xmin', [0], 'xmax', [0]);
+            testCase.verify_val(testCase.tri.mag_str, testCase.opt_tri_mag_str, ...
+                                'rel_tol', 1e-3);
+        end
+
+        function test_optmagstr_tri_af_custom_func_requires_xmin_and_xmax(testCase)
+            function [S, k, n] = custom_func(S0, x)
+                S = [1; 0; 0];
+                k = [1/3 1/3 0];
+                n = [0 0 1];
+
+            end
+            testCase.verifyError(@() testCase.tri.optmagstr('func', @custom_func), ...
+                                 'spinw:optmagtr:WrongInput');
+            testCase.verifyError(@() testCase.tri.optmagstr('func', @custom_func, 'xmin', [0]), ...
+                                 'spinw:optmagtr:WrongInput');
+            testCase.verifyError(@() testCase.tri.optmagstr('func', @custom_func, 'xmax', [0]), ...
+                                 'spinw:optmagtr:WrongInput');
+        end
+
+        function test_optmagstr_tri_af_epsilon(testCase)
+            % Test that large epsilon doesn't rotate spins
+            testCase.tri.optmagstr('epsilon', 1.);
+            expected_mag_str = testCase.opt_tri_mag_str;
+            expected_mag_str.k = [0 0 0]';
+            expected_mag_str.F = [0 0 1]';
+            testCase.verify_val(testCase.tri.mag_str, expected_mag_str, ...
+                                'rel_tol', 1e-3);
+        end
+
+        function test_optmagstr_tri_nRun(testCase)
+            % sw_timeit called in each nRun loop, and before and after
+            nRun = 4;
+            mock_sw_timeit = sw_tests.utilities.mock_function('sw_timeit');
+            testCase.tri.optmagstr(testCase.tri_optmagstr_args{:}, 'nRun', nRun);
+            testCase.assertEqual(mock_sw_timeit.n_calls, nRun + 2);
+            testCase.verify_val(testCase.tri.mag_str, testCase.opt_tri_mag_str, ...
+                                'rel_tol', 1e-3);
+        end
+
+        function test_optmagstr_tri_set_title(testCase)
+            title = 'Test';
+            out = testCase.tri.optmagstr(testCase.tri_optmagstr_args{:}, 'title', title);
+            testCase.assertEqual(out.title, title);
+            testCase.verify_val(testCase.tri.mag_str, testCase.opt_tri_mag_str, ...
+                                'rel_tol', 1e-3);
+        end
+
+        function test_optmagstr_tri_tid(testCase)
+            sw_timeit_mock = sw_tests.utilities.mock_function('sw_timeit');
+            tid = 2;
+            testCase.tri.optmagstr(testCase.tri_optmagstr_args{:}, 'tid', tid);
+            % check tid used in timing
+            for irow = 1:sw_timeit_mock.n_calls
+                testCase.assertEqual(sw_timeit_mock.arguments{irow}{3}, tid)
+            end
+            testCase.verify_val(testCase.tri.mag_str, testCase.opt_tri_mag_str, ...
+                                'rel_tol', 1e-3);
+        end
+
+        function test_optmagstr_optimisation_params(testCase)
```

Comment:
Done in [81b9205](https://github.com/SpinW/spinw/pull/117/commits/81b9205727f438f7fce8901c5dbda03cbdee703e). I found that you had to set both `'tolx'` and `'tolfun'` to stop the function converging, it would be much better if we could somehow mock the function but I think this will have to do.